cli/attach.go

```diff
@@ -0,0 +1,192 @@
+func (ac *AttachCommand) Init(c *Cli) {
+	ac.cli = c
+	ac.cmd = &cobra.Command{
+		Use:   "attach [OPTIONS] CONTAINER",
+		Short: "Attach local standard input, output, and error streams to a running container",
+		Long:  AttachDescription,
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return ac.runAttach(args)
+		},
+		Example: ac.example(),
+	}
+	ac.addFlags()
+}
+
+// addFlags adds flags for specific command.
+func (ac *AttachCommand) addFlags() {
+	flagSet := ac.cmd.Flags()
+	flagSet.BoolVar(&ac.noStdin, "no-stdin", false, "Do not attach STDIN")
+	flagSet.StringVar(&ac.detachKeys, "detach-keys", "", "Override the key sequence for detaching a container")
+	// TODO: sig-proxy will be supported in the future.
+	//flagSet.BoolVar(&ac.sigProxy, "sig-proxy", true, "Proxy all received signals to the process")
+}
+
+func inspectAndCheckState(ctx context.Context, cli client.CommonAPIClient, name string) (*types.ContainerJSON, error) {
+	c, err := cli.ContainerGet(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+	if !c.State.Running {
+		return nil, errors.New("You cannot attach to a stopped container, start it first")
+	}
+	if c.State.Paused {
+		return nil, errors.New("You cannot attach to a paused container, unpause it first")
+	}
+	if c.State.Restarting {
+		return nil, errors.New("You cannot attach to a restarting container, wait until it is running")
+	}
+
+	return c, nil
+}
+
+// runAttach is used to attach a container.
+func (ac *AttachCommand) runAttach(args []string) error {
+	name := args[0]
+
+	ctx := context.Background()
+	apiClient := ac.cli.Client()
+
+	c, err := inspectAndCheckState(ctx, apiClient, name)
+	if err != nil {
+		return err
+	}
+
+	if err := checkTty(!ac.noStdin, c.Config.Tty, os.Stdin.Fd()); err != nil {
+		return err
+	}
+
+	var inReader io.Reader = os.Stdin
+	if !ac.noStdin && c.Config.Tty {
+		in, out, err := setRawMode(!ac.noStdin, false)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "failed to set raw mode %s", err)
+			return fmt.Errorf("failed to set raw mode")
+		}
+		defer func() {
+			if err := restoreMode(in, out); err != nil {
+				fmt.Fprintf(os.Stderr, "failed to restore term mode %s", err)
+			}
+		}()
+
+		escapeKeys := defaultEscapeKeys
+		// Wrap the input to detect detach escape sequence.
+		// Use default escape keys if an invalid sequence is given.
+		if ac.detachKeys != "" {
+			customEscapeKeys, err := term.ToBytes(ac.detachKeys)
+			if err != nil {
+				return fmt.Errorf("Invalid detach keys (%s) provided", ac.detachKeys)
+			}
+			escapeKeys = customEscapeKeys
+		}
+		inReader = ioutils.NewReadCloserWrapper(term.NewEscapeProxy(os.Stdin, escapeKeys), os.Stdin.Close)
+	}
+
+	conn, br, err := apiClient.ContainerAttach(ctx, name, !ac.noStdin)
+	if err != nil {
+		return fmt.Errorf("failed to attach container: %v", err)
+	}
+	defer conn.Close()
+
+	// double check in case container in wrong state
+	_, err = inspectAndCheckState(ctx, apiClient, name)
```

Comment:
I don't think we need to check it again. For current IO design, the pipe copy goroutine will exit if the task has been stopped.

Reply:
ok, update